migration/migration_test.go

```diff
@@ -0,0 +1,434 @@
+func (ts testState) assertRecreatePVC(ctx context.Context, client kubernetesClient) error {
+	pvc, err := client.GetPVC(ctx, withMigrationSuffix(ts.pvc), ts.namespace)
+	if err != nil {
+		return err
+	}
+
+	if ann, ok := pvc.Annotations[isRenamedPVC]; !ok || ann != ts.pvc {
+		return errors.Errorf("annotation does not exist or is wrong: %v", pvc.Annotations[isRenamedPVC])
+	}
+
+	return makeDeletable(ctx, ts.pvc, ts.namespace, client)
+}
+
+// makeDeletable removes the finalizers from a PVC if it should still exist.
+func makeDeletable(ctx context.Context, pvcName, namespace string, client kubernetesClient) error {
+	pvc, err := client.GetPVC(ctx, pvcName, namespace)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrapf(err, "could not get PVC")
+	}
+	if !apierrors.IsNotFound(err) {
+		_, err = client.UpdatePVC(ctx, pvc, func(pvc *core.PersistentVolumeClaim) {
+			// remove the protection finalizer
+			pvc.Finalizers = nil
+		})
+		if err == nil {
+			return errors.Errorf("pvc still exists: %v", pvc.Finalizers)
+		}
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(err, "expected not found, got")
+		}
+	}
+	return nil
+}
+
+// assertSwitchTargetPVC sets the claimRef of the PV to the renamed PVC and updates the renamed
+// PVC's status.
+func (ts testState) assertSwitchTargetPVC(ctx context.Context, client kubernetesClient) error {
+	pv, err := client.GetPV(ctx, ts.pv)
+	if err != nil {
+		return err
+	}
+	if pv.Spec.ClaimRef != nil {
+		return errors.Errorf("PV still bound")
+	}
+
+	_, err = client.UpdatePV(ctx, pv, func(pv *core.PersistentVolume) {
+		pv.Spec.ClaimRef = &core.ObjectReference{
+			Namespace: ts.namespace,
+			Name:      withMigrationSuffix(ts.pvc),
+		}
+	})
+	if err != nil {
+		return err
+	}
+
+	pvc, err := client.GetPVC(ctx, withMigrationSuffix(ts.pvc), ts.namespace)
+	if err != nil {
+		return err
+	}
+	pvc.Status.Phase = core.ClaimBound
+
+	_, err = client.c.CoreV1().PersistentVolumeClaims(ts.namespace).UpdateStatus(ctx, pvc, meta.UpdateOptions{})
+	return err
+}
+
+// assertCreateTargetPVC ensures that the new PVC is created, has the correct storage class and
+// tthe argo annotations set.
+func (ts testState) assertCreateTargetPVC(ctx context.Context, client kubernetesClient) error {
+	pvc, err := client.GetPVC(ctx, ts.pvc, ts.namespace)
+	if err != nil {
+		return err
+	}
+
+	if pvc.Annotations["argocd.argoproj.io/sync-options"] != "Prune=false" {
+		return errors.Errorf("PVC does not have Argo Annotations")
+	}
+	if pvc.Spec.StorageClassName == nil {
+		return errors.Errorf("New PVC's storage Class is not set")
+	}
+
+	if *pvc.Spec.StorageClassName != testNewStorageClass {
+		return errors.Errorf("New PVC's storage Class is incorrect: %v", *pvc.Spec.StorageClassName)
+	}
+
+	return nil
+}
+
+// assertMigrateData ensures that a job is created, creates a dummy pod with that job as
+// an owner, sets the pod's status to succeeded and the job's status to succeeded too.
```

Comment:
I don't think you actually need to create the dummy pod here, the `job.Status.Succeeded = 1` should be enough, no?

Reply:
Technically, yes. But creating the pod here is used to assure that the `cleanup` stage works too & removes the pod again.
I'll add a comment to describe this.

Reply:
Sounds good :+1: